### DISCUSSION TITLE
Vehicle Routing Problem Pickup and Delivery in a certain order

### DISCUSSION THREAD

**Discussion by rahulj2409** [google/or-tools#2628] [Routing (and legacy CP) questions]
Python - Vehicle Routing Problem
I am trying to solve the Vehicle Routing Problem - Pickup and Delivery variant. A problem I am running into is that the pickup and delivery variant only seems to provide a "pickup needs to come before delivery, but there can be nodes visited in between". However I want to enforce an order on that. IE If "node A" is the pickup and "node B" is the dropoff, I want the vehicle to visit "node A", and then right after it, it visits "node B" in that specific order. Currently what the online example will do is something like "A -> D -> B" for instance, which I don't want. I would like to extend this to enforcing a certain order of the visiting of nodes for instance like if I say "visit A -> B-> C" in that order specifically.

I tried to treat the pickup and deliveries as one node and modify the distance matrix, but that won't work when I want to consider time windows for the pickup and deliveries.

**Comment by Letscho** [comment]
You could try to add constraints to the solver. e.g. you want A (index 1) to B (index 2) to C (index 3):
```python
routing.solver().Add(routing.NextVar(1) == 2)
routing.solver().Add(routing.NextVar(2) == 3)
```
This should force it to the order

If you run this example (https://developers.google.com/optimization/routing/pickup_delivery) and change the FirstSolutionStrategy to:

```python
search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)
```
the result will be:

---

Route for vehicle 0:
 0 ->  16 ->  14 ->  13 ->  12 -> 0
Distance of the route: 1780m

Route for vehicle 1:
 0 ->  7 ->  1 ->  6 ->  8 -> 0
Distance of the route: 1780m

Route for vehicle 2:
 0 ->  4 ->  3 ->  15 ->  11 -> 0
Distance of the route: 2032m

Route for vehicle 3:
 0 ->  5 ->  2 ->  10 ->  9 -> 0
Distance of the route: 1712m

Total Distance of all routes: 7304m

---

If you add the line 
```python
routing.solver().Add(routing.NextVar(13) == 15) 
```
it will force node 13 to be followed by 15 and the result will be:

---

Route for vehicle 0:
 0 ->  4 ->  3 -> 0
Distance of the route: 1392m

Route for vehicle 1:
 0 ->  7 ->  1 ->  6 ->  8 -> 0
Distance of the route: 1780m

Route for vehicle 2:
 0 ->  5 ->  2 ->  10 ->  16 ->  14 ->  9 -> 0
Distance of the route: 2192m

Route for vehicle 3:
 0 ->  13 ->  15 ->  11 ->  12 -> 0
Distance of the route: 1552m

Total Distance of all routes: 6916m

---

Be aware that the solution will take more time with this approach.

I hope this helps

**EDIT:**
This will only work if:

- The whole chain (A -> B -> C) is possible in the given assignment
- The whole chain is served by the same vehicle

Otherwise you would have to add extra constraints

---

This snippet ensures that each next node would be higher than the previous one, but only works if pickups/deliveries are ordered ([1,2], [3,4] ...). 
 
```python
for index in range(0, len(data['distance_matrix'])):
        nextIsHigher = routing.NextVar(index) >= (index + 1)
        nextIsDepot = routing.NextVar(index) == 0
        routing.solver().Add((nextIsHigher + nextIsDepot) == 1)
```

and produces a result similar to this:

---

Route for vehicle 0:
 0 ->  11 ->  12 ->  15 ->  16 -> 0
Distance of the route: 2464m

Route for vehicle 1:
 0 ->  1 ->  2 ->  5 ->  6 -> 0
Distance of the route: 2464m

Route for vehicle 2:
 0 ->  3 ->  4 ->  7 ->  8 -> 0
Distance of the route: 1848m

Route for vehicle 3:
 0 ->  9 ->  10 ->  13 ->  14 -> 0
Distance of the route: 2076m

Total Distance of all routes: 8852m
